staging/src/k8s.io/client-go/util/certificate/csr/csr.go

```diff
@@ -27,110 +27,248 @@ import (
+// PEM encoded CSR and send it to API server.
+func RequestCertificate(client clientset.Interface, csrData []byte, name string, signerName string, usages []certificatesv1.KeyUsage, privateKey interface{}) (reqName string, reqUID types.UID, err error) {
+	csr := &certificatesv1.CertificateSigningRequest{
 		// Username, UID, Groups will be injected by API server.
 		TypeMeta: metav1.TypeMeta{Kind: "CertificateSigningRequest"},
 		ObjectMeta: metav1.ObjectMeta{
 			Name: name,
 		},
-		Spec: certificates.CertificateSigningRequestSpec{
+		Spec: certificatesv1.CertificateSigningRequestSpec{
 			Request:    csrData,
 			Usages:     usages,
-			SignerName: &signerName,
+			SignerName: signerName,
 		},
 	}
 	if len(csr.Name) == 0 {
 		csr.GenerateName = "csr-"
 	}
 
-	req, err = client.Create(context.TODO(), csr, metav1.CreateOptions{})
+	reqName, reqUID, err = create(client, csr)
 	switch {
 	case err == nil:
+		return reqName, reqUID, err
+
 	case errors.IsAlreadyExists(err) && len(name) > 0:
 		klog.Infof("csr for this node already exists, reusing")
-		req, err = client.Get(context.TODO(), name, metav1.GetOptions{})
+		req, err := get(client, name)
 		if err != nil {
-			return nil, formatError("cannot retrieve certificate signing request: %v", err)
+			return "", "", formatError("cannot retrieve certificate signing request: %v", err)
 		}
 		if err := ensureCompatible(req, csr, privateKey); err != nil {
-			return nil, fmt.Errorf("retrieved csr is not compatible: %v", err)
+			return "", "", fmt.Errorf("retrieved csr is not compatible: %v", err)
 		}
 		klog.Infof("csr for this node is still valid")
+		return req.Name, req.UID, nil
+
 	default:
-		return nil, formatError("cannot create certificate signing request: %v", err)
+		return "", "", formatError("cannot create certificate signing request: %v", err)
 	}
-	return req, nil
 }
 
-// WaitForCertificate waits for a certificate to be issued until timeout, or returns an error.
-func WaitForCertificate(ctx context.Context, client certificatesclient.CertificateSigningRequestInterface, req *certificates.CertificateSigningRequest) (certData []byte, err error) {
-	fieldSelector := fields.OneTermEqualSelector("metadata.name", req.Name).String()
-	lw := &cache.ListWatch{
-		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-			options.FieldSelector = fieldSelector
-			return client.List(context.TODO(), options)
+func get(client clientset.Interface, name string) (*certificatesv1.CertificateSigningRequest, error) {
+	v1req, v1err := client.CertificatesV1().CertificateSigningRequests().Get(context.TODO(), name, metav1.GetOptions{})
+	if v1err == nil || !apierrors.IsNotFound(v1err) {
+		return v1req, v1err
+	}
+
+	v1beta1req, v1beta1err := client.CertificatesV1beta1().CertificateSigningRequests().Get(context.TODO(), name, metav1.GetOptions{})
+	if v1beta1err != nil {
+		return nil, v1beta1err
+	}
+
+	v1req = &certificatesv1.CertificateSigningRequest{
```

Comment:
I'm still reading, but a bootleg conversion like this is going to need a test to go with it.

Reply:
if we hit this path, we're talking to a pre-1.19 server, which wouldn't have the new fields :-)

Reply:
> if we hit this path, we're talking to a pre-1.19 server, which wouldn't have the new fields :-)

Sneaky.  I think technically I could turn off v1 csr intentionally, but I'll agree that's unlikely.

Reply:
> if we hit this path, we're talking to a pre-1.19 server, which wouldn't have the new fields :-)

That assumes defaults won't have value in this code.

Reply:
The same would apply to future fields added and defaulted in v1... if we start using new fields and depending on server-side defaults, we'll need a fallback when the field is absent from a v1 object read from a 1.19 server, not just in the conversion case